### PR TITLE
[tools] Replacement for danger

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Our danger.js bot doesn't work with PRs from forked repositories because GitHub doesn't pass secrets to such workflows, which seems right because otherwise someone could easily modify the workflow to steal the secrets. The solution would be to change workflow's `pull_request` event to `pull_request_target`, which executes the workflow on the target branch instead of PR's head (in this case GitHub passes the secrets to the workflow). However, danger doesn't work well with such event, and moreover, they don't recommend it so I think there is no chance they will fix it.

# How

- Added `expotools code-review` command that takes the PR, reviews its diff and posts warnings/errors as pull request review.
- Removed `danger.yml` workflow config, added `code-review.yml` instead.
- For now it only looks for missing changelog entries, but I'm gonna add much more things to review — including automatic code review suggestions (especially adding links to changelog entries).

# Test Plan

I've been running this command locally with `et review -p 12571` (see https://github.com/expo/expo/pull/12571#pullrequestreview-642743362)

# Considering to do later

- Check whether packages' `build` folder is up to date with source code (error, request for changes if isn't)
- Code review suggestions on changelog entries without links to the PR and author
- Post TypeScript linter issues
- Add GitHub labels such as `bot: approved`, `bot: rejected`
- Verify PR's description
- Remove `@expo/danger` package from the repository (let's keep it for a while, maybe we'll decide to reuse some of its code)

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]
LGTM 🚀🚀🚀

File: tools/src/code-review/reports.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * Generates the review comment based on given outputs and the commit the checks were run against.
+ */
+export function generateReviewBodyFromOutputs(
+  outputs: ReviewOutput[],
+  hasComments: boolean,
+  commitSha: string
+): string {
+  return [
+    "*Hi there! 👋 I'm a bot whose goal is to help you make your changes meet our guidelines.*",
```
Comment (cruzach) [inline tools/src/code-review/reports.ts:12]:
```suggestion
    "*Hi there! 👋 I'm a bot whose goal is to ensure your contributions meet our guidelines.*",
```

File: tools/src/code-review/reports.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ * Returns appropriate header for the review, depending on review results.
+ */
+function header(outputs: ReviewOutput[], hasComments: boolean): string {
+  if (outputs.length > 0) {
+    return `I've found some issues in your pull request that should be addressed (click on them to expand) 👇`;
+  }
+  if (hasComments) {
+    return `It's pretty good 👍 I just have a few comments to consider 👇`;
+  }
+  return 'Looks like I have nothing to complain about now 👏 Keep up the good work! 💪';
```
Comment (cruzach) [inline tools/src/code-review/reports.ts:45]:
```suggestion
  return 'Looks like I have nothing to complain about 👏 Keep up the good work! 💪';
```

File: tools/src/code-review/reviewers/checkMissingChangelogs.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  }
+
+  const changelogLinks = pkgsWithoutChangelogChanges
+    .map((pkg) => `- ${relativeChangelogPath(pullRequest, pkg)}`)
+    .join('\n');
+
+  return {
+    status: ReviewStatus.WARN,
+    title: 'Missing changelog entries',
+    body: `If you made some API or behavioral changes, please add appropriate entry to the following changelogs:
```
Comment (cruzach) [inline tools/src/code-review/reviewers/checkMissingChangelogs.ts:31]:
```suggestion
    body: `If you made some API or behavioral changes, please add an appropriate entry to the following changelogs:
```

File: tools/src/code-review/index.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+  // Run all the checks asynchronously and collects their outputs.
+  logger.info('🕵️‍♀️  Reviewing changes');
+  const outputs = (await Promise.all(REVIEWERS.map((reviewer) => reviewer(input)))).filter(
+    Boolean
+  ) as ReviewOutput[];
+
+  // Only active (non-passive) outputs will be reported in the review body.
+  const activeOutputs = outputs.filter(
+    (output) => output.title && output.body && output.status > ReviewStatus.PASSIVE
```
Comment (cruzach) [inline tools/src/code-review/index.ts:55]:
```suggestion
    (output) => output.title && output.body && output.status !== ReviewStatus.PASSIVE
```
personally i prefer `!==` here, as the significance of `> someEnum` isn't instantly understandable

File: tools/src/code-review/index.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  }
+}
+
+/**
+ * If any of the check failed, we want the review to request for changes.
+ * Otherwise, it's just a comment (and so fixes are not obligatory).
+ * There is no case where we approve the PR — we still want a human to review these changes :)
+ */
+function getReviewEventFromOutputs(outputs: ReviewOutput[]): GitHub.PullRequestReviewEvent {
+  return outputs.some((output) => output.status >= ReviewStatus.ERROR)
```
Comment (cruzach) [inline tools/src/code-review/index.ts:129]:
```suggestion
  return outputs.some((output) => output.status === ReviewStatus.ERROR)
```
same reasoning as other comment, unless we expect there to be more review statuses?

**Review by cruzach — COMMENTED** [review]
Very nice 👏 just a couple small change suggestions from me, excited to extend this further 😄

**Review by bbarthec — APPROVED** [review]
Excellent work 👍

**Pull request merged**